mgr/cephadm: force flag for ok-to-stop and ok-to-stop for monitoring stack

adk3798 opened the pull request:
Daemons that could cause data loss when stopped will always block.

Daemons that will only cause loss in availability should block but have
a workaround in the form of a force flag if the user is okay with the service
being down.

Also implements ok-to-stop for monitoring stack daemons that uses this system
of blocking on availability loss unless force flag is provided

Signed-off-by: Adam King <adking@redhat.com>
Signed-off-by: Daniel-Pivonka <dpivonka@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

sebastian-philipp commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -1416,12 +1416,19 @@ def _host_ok_to_stop(self, hostname: str) -> Tuple[int, str]:
             if dd.hostname == hostname:
                 daemon_map[dd.daemon_type].append(dd.daemon_id)
 
+        notifications = []  # type: List[str]
```
```suggestion
        notifications: List[str] = []
```

sebastian-philipp commented on src/pybind/mgr/cephadm/module.py:
```diff
@@ -1416,12 +1416,19 @@ def _host_ok_to_stop(self, hostname: str) -> Tuple[int, str]:
             if r.retval:
                 self.log.error(f'It is NOT safe to stop host {hostname}')
                 return r.retval, r.stderr
+            if r.stdout and 'NOTICE:' in r.stdout:
```
Would be great, if we could avoid introducing new, non-standard and undocumented serialization formats here in the mgr module.

jmolmo replied:
I think we can use directly any content in r.stdout not needed to check if it has a "NOTICE".

adk3798 replied:
I've changed it to just use all stdout. It includes the notifications and also messages about daemons that are considered okay to stop (whic I think is okay). Thoughts? @sebastian-philipp @jmolmo 
![image](https://user-images.githubusercontent.com/47704447/104335241-346e5780-54c1-11eb-8682-0bf4e5bc9a43.png)

sebastian-philipp replied:
The only think I'm worried about is to search for `'NOTICE:'` in the output of `ok_to_stop()` instead of providing a proper return value. Just imagine someone is fixing the error message to look like

`Notice, stopping the only daemon in the Prometheus service. ...`

And suddenly this code breaks. Without any warning or make check validation. And I don't want that. That's a nightmare for future refactorization.

adk3798 replied:
@sebastian-philipp would you prefer a designated return value for providing notifications, using any stdout that comes back (as in the example output) or a different approach entirely? I can see why searching the output string is an issue and will make sure to avoid that.

sebastian-philipp replied:
anything works for me.

jmolmo reviewed (approved):
For me is ok once solved the "serialization" problem pointed by Sebastian

tchaikov reviewed (changes requested):
could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes

mgfritch commented on src/pybind/mgr/cephadm/services/monitoring.py:
```diff
@@ -78,6 +80,19 @@ def config_dashboard(self, daemon_descrs: List[DaemonDescription]) -> None:
+        names = [f'{self.TYPE}.{d_id}' for d_id in daemon_ids]
+        out = f'It is presumed safe to stop {names}'
+        grafana_daemons = self.mgr.cache.get_daemons_by_type(self.TYPE)
+        if len(grafana_daemons) == 1:
+            # if only one grafana daemon, the one we are checking must be the only one
+            out = ('NOTICE: Stopping only daemon in Grafana service. Service will not be ' +
+                   'active while this daemon is stopped')
+        return HandleCommandResult(0, out, None)
```
the ok-to-stop commands for osd, mon, mds types require availability (e.g. cannot stop all mds nor reduce the count below 1) ...

should the same apply here? maybe we should return an EBUSY but allow for a flag to ignore the ok-to-stop failures during host maintenance?

sebastian-philipp reviewed (commented):
That's much better! Do you think it's easy to reduce the amount of copy&paste?

adk3798 commented:
Just adjusted this pretty heavily to add a force flag to ok-to-stop rather than just adding notice messages for monitoring daemons.

Things to note:

- changed _host_ok_to_stop to run through all daemons rather than returning on first failure. The idea here being if a user wanted to put a host in maintenance that had two daemons that would cause availability issues we want to show both of them rather than just showing one and then having a subsequent call with the force flag override both.
- Did not include full use of this flag in places other than host maintenance. `orch stop` and `orch daemon stop` should likely make use of ok-to-stop and therefore this flag. I've left them alone for now because the way daemon actions is implemented through scheduling involves storing the action in a [host][daemon] -> action (str) mapping. I think that data structure would have to be adjusted (and therefore everywhere that uses the scheduled daemon actions as well)  to pass the information that the force  flag was set along. I could include that in this PR but I think it could also be its own change.
- applying service specs also makes use of ok-to-stop. I currently have it default to force=True there as that maintains the old functionality. The apply command could theoretically also make use of the force flag. Again, not sure if we want something like that in this PR.
- Currently, all the monitoring stack ok-to-stop functions just use an error code of 1 if they're going to block. I'm open to suggestions for better values to use.

jmolmo commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -207,7 +207,7 @@ def _check_and_set_dashboard(self,
                 logger.warning('Failed to set Dashboard config for %s: %s', service_name, e)
 
-    def ok_to_stop(self, daemon_ids: List[str]) -> HandleCommandResult:
+    def ok_to_stop(self, daemon_ids: List[str], force: Optional[bool] = False) -> HandleCommandResult:
```
I suggest to add a general use method in CephadmService class to avoid repeat things when we want to check  the "ok_to_stop" in different services:

for example:

```
    def is_safe_ok_to_stop(self, daemon_type: str, daemon_ids: List[str], service: str, low_limit: int) -> Tuple[bool, str]:
        # Provides a warning about if it possible or not to stop <n> daemons in a service
        names = [f'{daemon_type}.{d_id}' for d_id in daemon_ids]
        out = f'It is presumed safe to stop {names}'
        warn = False
        number_of_daemons = len(self.mgr.cache.get_daemons_by_type(daemon_type))
        if number_of_daemons < low_limit:
            warn = True
            daemon_str = 'only daemon' if number_of_daemons == 1 else f'{number_of_daemons} daemons'
            daemons_left_word = 'daemon' if number_of_daemons-low_limit == 1 else 'daemons'
            daemons_left_str = 'no daemons' if number_of_daemons-low_limit == 0 else f'{number_of_daemons-low_limit} {daemons_left_word}'
            low_limit_daemons_str = f'{low_limit} daemon' if low_limit == 1 else f'{low_limit} daemons'
            out = (f'WARNING: Stopping {daemon_str} in {service} service.'
                    'Service will not be operational with {daemons_left_str}.'
                    'At least {low_limit_daemons_str} must be running to guarantee service')
        return (warn, out)
```

jmolmo commented on src/pybind/mgr/cephadm/services/monitoring.py:
```diff
@@ -78,6 +80,20 @@ def config_dashboard(self, daemon_descrs: List[DaemonDescription]) -> None:
             service_url
         )
 
+    def ok_to_stop(self, daemon_ids: List[str], force: Optional[bool] = False) -> HandleCommandResult:
```
Using the proposed "is_safe_ok_to_stop"  CephadmService  method we can simplify the ok_to_stop methods for specific services, for example for Grafana (at least 1 daemon needed to provide service)

```python
        warn, warn_message = self.is_safe_ok_to_stop(self.TYPE, daemon_ids, 'Grafana', 1)
        if warn and not force:
            return HandleCommandResult(1, None, warn_message)
        return HandleCommandResult(0, warn_message, None)
```

sebastian-philipp commented:
As a general rule of thumb: Please don't use `Optional[bool]` for parameter annotations. 

The difference is: 

```python
def my_func(param: Optionl[bool]): ...
```
declares a variable as nullable. If you use `Optionl[bool]`, param can be either `True` or `False` or `None` and you have to check that in the code. `param = True` or `False` is not identical to `param = True` or `False` or `None`.  And on the other hand:

```python
def my_func(param: bool = False): ...
```

declares `param` as optional when calling the function.

cc @ceph/orchestrators

sebastian-philipp commented:
https://pulpito.ceph.com/swagner-2021-01-15_09:42:49-rados:cephadm-wip-swagner-testing-2021-01-14-1551-distro-basic-smithi/

* https://tracker.ceph.com/issues/48894
* https://tracker.ceph.com/issues/48715
* https://tracker.ceph.com/issues/48142

sebastian-philipp commented:
https://pulpito.ceph.com/swagner-2021-01-18_11:41:36-rados:cephadm-wip-swagner-testing-2021-01-15-1448-distro-basic-smithi/

Failures

https://tracker.ceph.com/issues/48894
https://tracker.ceph.com/issues/48715
https://tracker.ceph.com/issues/48142

sebastian-philipp commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -230,6 +230,24 @@ def ok_to_stop(self, daemon_ids: List[str]) -> HandleCommandResult:
         logger.info(out)
         return HandleCommandResult(r.retval, out, r.stderr)
 
+    def _enough_daemons_to_stop(self, daemon_type: str, daemon_ids: List[str], service: str, low_limit: int) -> Tuple[bool, str]:
+        # Provides a warning about if it possible or not to stop <n> daemons in a service
+        names = [f'{daemon_type}.{d_id}' for d_id in daemon_ids]
+        out = f'It is presumed safe to stop {names}'
+        warn = False
+        number_of_daemons = len(self.mgr.cache.get_daemons_by_type(daemon_type))
+        if number_of_daemons <= low_limit:
+            warn = True
+            daemon_str = 'only daemon' if number_of_daemons == 1 else f'{number_of_daemons} daemons'
+            daemons_left_word = 'daemon' if number_of_daemons - low_limit == 1 else 'daemons'
+            daemons_left_str = 'no daemons' if number_of_daemons - \
+                low_limit == 0 else f'{str(number_of_daemons - low_limit)} {daemons_left_word}'
+            low_limit_daemons_str = f'{low_limit} daemon' if low_limit == 1 else f'{low_limit} daemons'
+            out = (f'WARNING: Stopping {daemon_str} in {service} service. '
+                   f'Service will not be operational with {daemons_left_str}. '
+                   f'At least {low_limit_daemons_str} must be running to guarantee service. ')
+        return (warn, out)
```
had a hard time understanding this block. Maybe use something like 

```python
    def _enough_daemons_to_stop(self, daemon_type: str, daemon_ids: List[str], service: str, low_limit: int) -> Tuple[bool, str]:
        # Provides a warning about if it possible or not to stop <n> daemons in a service
        names = [f'{daemon_type}.{d_id}' for d_id in daemon_ids]
        number_of_daemons = len(self.mgr.cache.get_daemons_by_type(daemon_type))
        if number_of_daemons > low_limit:
            return False, f'It is presumed safe to stop {names}'
        
        num_daemons_left = number_of_daemons - low_limit
        
        def plural(word, plural_char, count):
            return word if count == 1 else f'{word}{plural_char}'
        
        daemon_count = "only" if number_of_daemons == 1 else number_of_daemons
        left_count = "no" if num_daemons_left == 0 else num_daemons_left

        out = (f'WARNING: Stopping {daemon_count} {plural("daemon", "s", number_of_daemons)} in {service} service.'
                f'Service will not be operational with {left_count} {plural("daemon", "s", low_limit)}. '
                f'At least {low_limit} {plural("daemon", "s", low_limit)} must be running to guarantee service. ')
        return (True, out)
``` 
instead??

sebastian-philipp commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -8,6 +8,7 @@
 from cephadm.serve import CephadmServe
 from cephadm.utils import name_to_config_section, CEPH_UPGRADE_ORDER
 from orchestrator import OrchestratorError, DaemonDescription
+from orchestrator._interface import daemon_type_to_service, service_to_daemon_types
```
```suggestion
from orchestrator import daemon_type_to_service, service_to_daemon_types
```

Please don't import anything from `_interface` as this is a private python module. See

https://github.com/ceph/ceph/blob/3dd11ec576a99fb0452a4760e56b790693d7a18b/src/pybind/mgr/orchestrator/__init__.py#L10-L19

sebastian-philipp commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -177,7 +178,9 @@ def _wait_for_ok_to_stop(self, s: DaemonDescription) -> bool:
                 return False
 
-            r = self.mgr.cephadm_services[s.daemon_type].ok_to_stop([s.daemon_id])
+            # setting force flag to retain old functionality. Should revisit.
```
If we set `force=False` for small clusters, we might block the upgrade due to not enough grafana instances. I think keeping `force=True` here is the right thing to do.

sebastian-philipp commented on src/pybind/mgr/orchestrator/_interface.py:
```diff
@@ -829,7 +829,7 @@ def host_ok_to_stop(self, hostname: str) -> Completion:
         raise NotImplementedError()
 
-    def enter_host_maintenance(self, hostname: str) -> Completion:
+    def enter_host_maintenance(self, hostname: str, force: bool = False) -> Completion:
```
mind to document the force parameter?

sebastian-philipp commented on src/pybind/mgr/orchestrator/module.py:
```diff
@@ -393,11 +393,11 @@ def _host_ok_to_stop(self, hostname: str):
     @_cli_write_command(
         'orch host maintenance enter')
-    def _host_maintenance_enter(self, hostname: str):
+    def _host_maintenance_enter(self, hostname: str, force: bool = False):
         """
         Prepare a host for maintenance by shutting down and disabling all Ceph daemons (cephadm only)
         """
```
mind to document the force parameter?

sebastian-philipp commented:
queued: https://pulpito.ceph.com/swagner-2021-01-26_11:54:26-rados:cephadm-wip-swagner-testing-2021-01-25-1228-distro-basic-smithi/

sebastian-philipp commented:
https://pulpito.ceph.com/swagner-2021-01-26_11:54:26-rados:cephadm-wip-swagner-testing-2021-01-25-1228-distro-basic-smithi/

lots of podman failures due to https://github.com/containers/podman/issues/9096
